### PR TITLE
Feature/add utils

### PR DESCRIPTION
**Pull request by KGSDilshan**
Doesn't have a release, just code

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -16,4 +25,11 @@ export default {
         validateBody,
         validateQuery,
     },
+    sleep,
+    convertTimeStringToSeconds,
+    randomElement,
+    roundToNearest,
+    deleteFiles,
+    decapitalizeFirstLetter,
+    randomRange,
```
Comment (KGS-Ben) [inline index.ts:34]: Comment:

I'd prefer if these were nested into their respective 'subModule'/folder name, similar to `validateBody` being put into `expressMiddlewares`. Declutters the namespace IMO.

```
const { AxiosDecorator, Lru } = require('@kgs-research/utils');
const { sleep, convertTimeStringToSeconds } = require('@kgs-research/utils').generic;
```

vs

```
const { AxiosDecorator, Lru, sleep, convertTimeStringToSeconds } = require('@kgs-research/utils');
```

File: lib/generic/index.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ */
+import fs from 'fs/promises';
+
+/**
+ * Asynchronous sleep function
+ *
+ * @param {number} millis - time to sleep (ms)
+ * @returns {Promise<void>} - Promise that resolves after millis
+ */
+export async function sleep(millis: number): Promise<void> {
```
Comment (KGS-Ben) [inline lib/generic/index.ts:12]: Update:

This shouldn't be an async function.

File: lib/generic/index.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    return new Promise(resolve => setTimeout(resolve, millis));
+}
+
+/**
+ * Convert a time string like '17 min 30 sec' to seconds
+ *
+ * @param {String} timeString Time string
+ * @returns {Number} Time in seconds
```
Comment (KGS-Ben) [inline lib/generic/index.ts:20]: Update:
Hyphens b/w param name and description

File: lib/generic/index.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+ */
+export function roundToNearest(toRound: number, digits: number = 2): number {
+    let tens = Math.pow(10, digits) || 1;
+    return Math.round(toRound * tens) / tens;
+}
+
+/**
+ * Delete any number of files.
+ *
+ * @param {...string} filesToDelete - List of files to delete
```
Comment (KGS-Ben) [inline lib/generic/index.ts:67]: Update:

Add `@throws` to docstring

File: lib/generic/index.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            throw new Error(`Failed to delete file ${filename}\n${err}`);
+        }
+    }
+}
+
+/**
+ * Decapitalize first character of a string
+ *
+ * @param {string} str - String to decap
+ * @returns {string} Decapped string
```
Comment (KGS-Ben) [inline lib/generic/index.ts:84]: Update:

Un-abbreviate decap

File: lib/generic/index.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+ */
+export function decapitalizeFirstLetter(str: string): string {
```
Comment (KGS-Ben) [inline lib/generic/index.ts:86]: Nit:
Should we handle undefined/null strings? It's been a pretty common unhandled error.

File: lib/generic/index.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    return str.charAt(0).toLowerCase() + str.slice(1);
+}
+
+/**
+ * Generate a random number within a range.
+ *
+ * @param {number} min Lower bound of random number
+ * @param {number} max Upper bound of random number
```
Comment (KGS-Ben) [inline lib/generic/index.ts:94]: Update:
Hyphens b/w param and description

**Review by KGS-Ben — CHANGES_REQUESTED** [review]
Need to update the version number since this will add new features and needs to be published to npm

File: lib/generic/index.ts (hunk 2 above)
Comment (KGSDilshan) [reply lib/generic/index.ts:12]: what do you mean? It's meant to be used like:
```
loop:
   someFunctionCalls();
   await sleep(n);
```
The caller uses it as asyc

Comment (KGS-Ben) [reply lib/generic/index.ts:12]: The actual function (sleep) is not async, but it's returned value is.

Right now it's a function that returns a promise containing a promise (`Promise<Promise<void>>` vs `Promise<void>`)

**Comment by KGS-Ben** [comment]
> Need to update the version number since this will add new features and needs to be published to npm

Still needs to be addressed

**Comment by KGSDilshan** [comment]
> > Need to update the version number since this will add new features and needs to be published to npm
> 
> Still needs to be addressed

Shouldn't that be a separate issue? Pipeline would be:
1) Dev code and introduce to package
2) Bundle features in release
3) Release

I think we should release on a separate PR than the PR for code commits

**Comment by KGS-Ben** [comment]
> > > Need to update the version number since this will add new features and needs to be published to npm
> > 
> > 
> > Still needs to be addressed
> 
> Shouldn't that be a separate issue? Pipeline would be:
> 
> 1. Dev code and introduce to package
> 2. Bundle features in release
> 3. Release
> 
> I think we should release on a separate PR than the PR for code commits

Sure. I await the release PR.

**Pull request merged**